base/promotion.jl

```diff
@@ -217,13 +217,38 @@ max(x::Real, y::Real) = max(promote(x,y)...)
 min(x::Real, y::Real) = min(promote(x,y)...)
 minmax(x::Real, y::Real) = minmax(promote(x, y)...)
 
-# "Promotion" that takes a function into account. You can override this
-# as needed. For example, if you need to provide a custom result type
-# for the multiplication of two types,
-#   promote_op{R<:MyType,S<:MyType}(::typeof(*), ::Type{R}, ::Type{S}) = MyType{multype(R,S)}
-promote_op(::Any)    = (@_pure_meta; Bottom)
-promote_op(::Any, ::Any, ::Any...) = (@_pure_meta; Any)
-promote_op{T}(::Type{T}, ::Any) = (@_pure_meta; T)
+# "Promotion" that takes a function into account. These are meant to be
+# used mainly by broadcast methods, so it is adviced against overriding them
+if isdefined(Core, :Inference)
+    function _promote_op(op, T::Type)
+        G = Tuple{Generator{Tuple{T},typeof(op)}}
+        R = Core.Inference.return_type(first, G)
```

Comment:
That was the first thing I tried, the problem is that Test.@inferred does not work that way, but it works like this.